doc/show.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SHOW" "1" "December 2018" "" ""
.
.SH "NAME"
\fBshow\fR \- An interactive directory/file browser written for Unix\-like systems\.
.
.SH "USAGE"
\fBshow\fR \fI<options\fR> [\fItarget\fR]
.
.SH "DESCRIPTION"
DF\-SHOW: An interactive directory/file browser written for Unix\-like systems\. Based on the applications from the PC\-DOS DF\-EDIT suite by Larry Kroeker\.
.
.SH "OPTIONS"
The following options are shared with \fBls\fR:
.
.TP
\fB\-a\fR, \fB\-\-all\fR
Does not ignore files or directories starting with \fB\.\fR\. The \fB\.\fR and \fB\.\.\fR objects are never hidden due to their navigational functions in \fBshow\fR\.
.
.TP
\fB\-\-author\fR
Prints the author of the file in the output\.
.
.TP
\fB\-B\fR, \fB\-\-ignore\-backups\fR
Hides files ending in \fB~\fR, often used as backup files\.
.
.TP
\fB\-\-color\fR[=\fICOLOR\fR]
Sets colorizing of files and directories\. Can be set to \fBalways\fR or \fBnever\fR\.
.
.TP
\fB\-f\fR
Does not sort the output, this is equivilent to running \fBshow\fR with \fB\-aU\fR\.
.
.TP
\fB\-g\fR
Only displays the group of each object\. Can be paird with \fB\-\-author\fR\.
.
.TP
\fB\-G\fR, \fB\-\-no\-group\fR
Omits the group of each object in the output\.
.
.TP
\fB\-h\fR, \fB\-\-human\-readable\fR
Displays the file sizes in a human readable format\. For example, 1024 Bytes will be represented as 1M\.
.
.TP
\fB\-\-si\fR
As above, but uses powers of 1000\. (1000 = 1m)
.
.TP
\fB\-r\fR, \fB\-\-reverse\fR
Reverse the order of sorting\. Can be paird with \fB\-S\fR or \fB\-t\fR
.
.TP
\fB\-S\fR
Sorts files and directories by size, largest first\.
.
.TP
\fB\-\-time\-style\fR=[\fITIME_STYLE\fR]
Time/Date format\. Can be set to \fBfull\-iso\fR, \fBlong\-iso\fR, \fBiso\fR, or \fBlocale\fR\.
.
.TP
\fB\-t\fR
Sorts by modification time, newset first\.
.
.TP
\fB\-U\fR
Displays output in the order they\'re found\.
.
.TP
\fB\-\-help\fR
Displays help message, then exits\.
.
.TP
\fB\-\-version\fR
Displays program version, then exits\.
.
.P
The following options are specific to \fBshow\fR:
.
.TP
\fB\-\-theme\fR=[\fITHEME\fR]
Color themes\. Passing this argument without an option will display available themes\.
.
.TP
\fB\-\-no\-danger\fR
Disables the danger color scheme when running with elevated privileges\.
.
.TP
\fB\-\-marked\fR=[\fIMARKED\fR]
Displays information about marked files\. Can be set to \fBalways\fR, \fBnever\fR, or \fBauto\fR\.
.
.TP
\fB\-\-no\-sf\fR
Doesn\'t display files using \fBsf\fR (if installed) and defaults to the user\'s PAGER setting\.
.
.SH "OVERVIEW"
The \fBshow\fR application lets users view the names of files and directories on a disk with information about the files\. Files can be copied, moved, viewed, and edited (in your system\'s default editor)\. The application is run using the show command\. The output is similar to the \fBls\fR command\.
.
.P
The screen is devided into three parts\. The first is the command bar which displays the functions available, their hotkeys, and aditional prompts when requested\. The commands are described in more detail in the \fICOMMANDS\fR and \fIGLOBAL COMMANDS\fR sections\.
.
.P
The second part is the middle of the screen between the top and bottom rows\. This displays information for the files and directories and is known as the display window\. The top three rows of the display window are header lines which identify the information being displayed in each column, the number of objects in the current directory, the total space used by the files in the current directory, the remaining disk space on the mounted partition the directory resides, and the current path\. Each line gives the object\'s permissions, hard links, the owner and group information, size of the object, date and time of the last modification, as well as the name of each file or directory\.
.
.P
The third part is known as the function bar, each function can be accessed by its numeric function key\. The functions are described in more deatail in the \fIFUNCTIONS\fR section\.
.
.SH "COMMANDS"
Several commands are provided which can be used on a single file or multiple files\. These commands provide a lot of the functions supported by standard Unix commands, but in a convenient user interface\. The file to be operated on is selected by navigating the cursor on the line of the desired file\. The highlighed (first character) of the command is used to invoke its function\. Multiple files can be selected using the "Mark/Unmark" function\. The following is a list of valid commands and a brief description of each\.
.
.TP
\fBCopy\fR
Copy file or marked files to a specified destination\.
.
.TP
\fBDelete\fR
Delete file or marked files\.
.
.TP
\fBEdit\fR
Invoke the system\'s text editor\.
.
.TP
\fBHidden\fR
Toggle the visibility of hidden objects\.
.
.TP
\fBModify\fR
Change the owner/group or permissions of files\.
.
.TP
\fBQuit\fR
Quit current directory\.
.
.TP
\fBRename\fR
Rename/move file or marked files\.
.
.TP
\fBShow\fR
Show file or directory\.
.
.TP
\fBeXec\fR
Runs files marked as executable\.
.
.TP
\fBhUnt\fR
Searches and marks files matching a string\.
.
.P
Each of the commands are described in the following paragraphs\. If the command has any additional propts, they are also described here\.
.
.TP
\fBCopy\fR
Copy the selected file, or marked files to another file location\. When using this command, only the destination must be specified\. The destination file name is requested by one of the following prompts\. An empty response to the prompt or the \fIESC\fR key is used to abort the command\. If there are no marked files, a single file copy is requested; otherwise, a multiple file copy is requested\.
.
.IP
\fBCopy file to:\fR
.
.IP
\fBCopy multiple files to:\fR
.
.IP
If the destination file already exists, confirmation to replace the file is requested by the following prompt\.
.
.IP
\fBReplace file [<file name>]? (Yes/No)\fR
.
.TP
\fBDelete\fR
Delete the selected file, or marked files\. As this is a destructive process, confirmation is requested by one of the following prompts\. If there are no marked files, a single file delete is requested; otherwise, a multiple delete is requested\.
.
.IP
\fBDelete file? (Yes/No)\fR
.
.IP
\fBDelete file [<file name>]? (Yes/No/All/Stop)\fR
.
.IP
For single files, a "Y" will delete files and anything else will abort the delete file operation\. For multiple files, confirmation for all files is requested\. A "Y" will delete the prompted file, an "N" will not delete the file, an "A" will delete all the marked files without further confirmation, and an "S" will stop the multiple delete command\.
.
.TP
\fBEdit\fR
This invokes the default text editor\. Please consult the \fBman\fR pages of your preferred editor for more information\. The edit command utilizes the \fBVISUAL\fR or \fBEDITOR\fR environment variable\. If this has not been set, the following message is displayed\.
.
.IP
\fBPlease export a VISUAL environment variable to define the utility program name\.\fR
.
.IP
To resolve this, set a variable to your preferred editor in your shell\'s user preference file\.
.
.IP
\fBexport VISUAL=/usr/bin/vi\fR
.
.TP
\fBHidden\fR
This toggles the display of hidden files and directories\. This allows adjustment of the \fB\-a\fR option whilst in the application\. If the currently selected object is hidden by this toggle, the cursor is returned to the top of the directory\.
.
.TP
\fBModify\fR
Modify the owner/group properties or changes the permissions of an object, or multiple objects\. When invoked, the following prompt is displayed\.
.
.IP
\fBModify: Owner/Group, Permissions\fR
.
.IP
Selecting "O" will bring up two prompts\.
.
.IP
\fBSet Owner:\fR
.
.IP
\fBSet Group:\fR
.
.IP
If the names of the owner or group is invalid, an error will be displayed to the user\.
.
.IP
Selecting "P" will bring up the following prompt\.
.
.IP
\fBModify Permissions:\fR
.
.IP
The syntax is a 3 or 4 digit octect\. See the \fBman\fR pages for \fBchmod\fR for further information\.
.
.TP
\fBQuit\fR
Closes the current directory currently displayed\. The application will return you to the previous directory you were viewing\. If there are no previous directory, a blank screen showing the global commands is displayed\.
.
.TP
\fBRename\fR
Rename the selected file, or marked files to a new file name\. The new file name is requested by one of the following prompts\. After a file is renamed, the new file information line is into the list (providing the file has been renamed to the same directory) and the old information line is removed\. Files can only be moved to locations on the same mounted partition\. An empty response to the prompt or the \fIESC\fR key will abort the command\. If there are no marked files, a single file rename is requested; otherwise a multiple file rename is requested\.
.
.IP
\fBRename file to:\fR
.
.IP
\fBRename multiple files to:\fR
.
.TP
\fBShow\fR
Will either display the contents of a directory, or open the contents of a file\. The show file command utilizes the \fBPAGER\fR environment variable\. If this has not been set, the following message is displayed\.
.
.IP
\fBPlease export a PAGER environment variable to define the utility program name\.\fR
.
.IP
To resolve this, set a variable to your preferred pager in your shell\'s user preference file\.
.
.IP
\fBexport PAGER=/usr/bin/less\fR
.
.TP
\fBeXec\fR
Will execute the selected file if it has the execute flag set and the current user running the \fBshow\fR process has permission to\. Arguments are requested by the following prompt\. Unlike other commands, an empty response will execute the file without arguments\. To abort at this prompt, the \fIESC\fR key must be used\.
.
.IP
\fBArgs to pass to <file>:\fR
.
.IP
The following error is displayed if the file does not have an executable flag, or the user does not have sufficient privileges to run\.
.
.IP
\fBError: Permission denied\fR
.
.TP
\fBhUnt\fR
Hunts the selected file, or marked files containing a regex string\. When used with a single file, the selected file will be marked if the string matches\. When using multiple files, any files not matching the string will be unselected\. Case sensitivity is requested by the following prompt, afterwards the user is asked to input the string to search\. To abort at this prompt, the \fIESC\fR key must be used\.
.
.IP
\fBCase Sensitive, Yes/No/ESC (enter = no)\fR
.
.IP
Following this selection, one of the following prompts will be displayed\.
.
.IP
\fBMatch Case \- Enter string:\fR
.
.IP
\fBIgnore Case \- Enter string:\fR
.
.SH "FUNCTIONS"
In order to select objects to be used by the commands described in the previous section, the cursor must be moved to the line of the desired object\. The functions to move the cursor and the list of files in the display window are described here\. A list of the valid functions and their associated function keys is given list\.
.
.TP
\fBF1\fR, \fBPgDn\fR
Page Down
.
.TP
\fBF2\fR, \fBPgUp\fR
Page Up
.
.TP
\fBF3\fR
Top of List
.
.TP
\fBF4\fR
Bottom of List
.
.TP
\fBF5\fR
Refresh Directory
.
.TP
\fBF6\fR
Mark/Unmark File
.
.TP
\fBF7\fR
Mark All Files
.
.TP
\fBF8\fR
Unmark All Files
.
.TP
\fBF9\fR
Sort List
.
.TP
\fBF10\fR
Block Mark
.
.TP
\fBHOME\fR
Top of Display
.
.TP
\fBEND\fR
Bottom of Display
.
.TP
\fBDown\fR, \fBRETURN\fR
Down one line
.
.TP
\fBUp\fR
Up one line
.
.TP
\fBRight\fR
Right one column
.
.TP
\fBLeft\fR
Left one column
.
.TP
\fBESC\fR
Global Commands
.
.P
The display functions with their associated key assignments are described here\.
.
.TP
\fBPage Down\fR
\fBF1, PgDn\fR: Scroll the display window down or forward a page on the list of files\. The cursor is left in the same relative row of the window unless the end of the list is reached\. If the last file of the list is already displayed in the window, the list is not scrolled, but the cursor is placed on the last file in the list\.
.
.TP
\fBPage Up\fR
\fBF2, PgUp\fR: Scroll the display window up or backward a page on the list of files\. The cursor is left in the same relative row of the window unless the beginning of the list is reached\. If the first file of the list is already displayed in the window, the list is not scrolled, but the cursor is placed on the first file in the list\.
.
.TP
\fBTop of List\fR
\fBF3\fR: Display the beginning of the list of files in the window and place the cursor on the first file in the list\.
.
.TP
\fBBottom of List\fR
\fBF4\fR: Display the end of the list of files in the display window and place the cursor on the last file of the list\.
.
.TP
\fBRefresh Directory\fR
\fBF5\fR: Rereads the directory\. This function is useful to update the list of files after several new files have been created or updated outside of the application\.
.
.TP
\fBMark/Unmark File\fR
\fBF6\fR: Toggle the file mark on the current file\. The file mark is indicated with an "*" in front of the file name\.
.
.TP
\fBMark All Files\fR
\fBF7\fR: Set the file mark on all the files but not directories in the list\.
.
.TP
\fBUnmark All Files\fR
\fBF8\fR: Remove the file marks from all files in the list\.
.
.TP
\fBSort List\fR
\fBF9\fR: Normally, the file list is sorted by file name alphabetically\. This function allows the files to be listed based on another sorting criteria which is requested by the following promt\.
.
.IP
\fBSort list by \- Date & time, Name, Size\fR
.
.IP
The option is selected by using the first letter of the option name\.
.
.IP
\fBDate & time\fR: Sort the list on date and time so the newest files are at the top of the list\.
.
.IP
\fBName\fR: Sort the list on the file name\.
.
.IP
\fBSize\fR: Sort the list on file size so the largest are at the top of the list\.
.
.IP
Using \fISHIFT\fR whilst selecting an option performs that action in reverse order\.
.
.TP
\fBBlock Mark\fR
\fBF10\fR: Marks all files between two points\. Files marked will be indicated with an \fB*\fR in front of them\.
.
.TP
\fBTop of Display\fR
\fBHOME\fR: Move the cursor to the first file on the current display\.
.
.TP
\fBBottom of Display\fR
\fBEND\fR: Move the cursor to the last file on the current display\.
.
.TP
\fBDown One Line\fR
\fBDown Arrow, Return\fR: Move the cursor down one line to the next file in the display\. If the cursor is on the bottom row of the window, the window is scrolled down one line\. If the present line is the last file in the list, the cursor is not repositioned\.
.
.TP
\fBUp One Line\fR
\fBUp Arrow\fR: Move the cursor up one line to the next file in the display\. If the cursor is on the top row of the window, the window is scrolled up one line\. If the present line is the first file in the list, the cursor is not repositioned\.
.
.TP
\fBRight one column\fR
\fBRight Arrow\fR: Moves the display area one column\. This occurs when an entry rolls off the edge of the display\. Scrolling will stop at the end of the longest entry\.
.
.TP
\fBLeft one column\fR
\fBLeft Arrow\fR: Moves the display area one column\.
.
.TP
\fBGlobal Commands\fR
\fBESC\fR: Invoke the \fIGLOBAL COMMANDS\fR described in the next section\. This allows another directory to be displayed without terminating the current display\.
.
.SH "GLOBAL COMMANDS"
When a file group display is terminated with the Quit command, one of the following commands can be used to display another group of files, invoke the editor for a fire, or terminate the application completely\. The first character of the command is used to invoke the desired function\. The command line is shown below\.
.
.P
\fBcOlors, Run command, Edit file, Help, Make dir, Quit, Show dir\fR
.
.P
These commands are desctibed below\.
.
.TP
\fBcOlors\fR
Launches an inbuilt color configuration utility which cusomizes the colors for the various display areas in all the utilities\. Further information can be found in the \fICOLORS\fR section\.
.
.TP
\fBRun command\fR
Invoke your shell\. The \fBshow\fR application is still resident, so the "exit" command will return to the application\.
.
.TP
\fBEdit file\fR
Invoke the default text editor to edit the specified file\. The file name is requested by the following prompt\. An empty response is used to abort this command\.
.
.IP
\fBEdit File \- Enter pathname:\fR
.
.TP
\fBHelp\fR
Launches the \fBman\fR pages for \fBshow\fR\.
.
.TP
\fBMake dir\fR
Make a new directory\. The directory name is requested by the following prompt\. An empty response is used to abort this command\.
.
.IP
\fBMake Directory \- Enter pathname:\fR
.
.TP
\fBQuit\fR
Terminate \fBshow\fR\.
.
.TP
\fBShow dir\fR
Invoke the application to display another directory\. The directory name is requrested by the following prompt\. An empty response is used to abort this command\.
.
.IP
\fBShow Directory \- Enter pathname:\fR
.
.SH "COLORS"

src/show.c
/*
  DF-SHOW - A clone of 'SHOW' directory browser from DF-EDIT by Larry Kroeker
  Copyright (C) 2018  Robert Ian Hawdon

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <ncurses.h>
#include <locale.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <getopt.h>
#include <libconfig.h>
#include "config.h"
#include "showfunctions.h"
#include "showmenus.h"
#include "colors.h"
#include "common.h"
#include "show.h"

char currentpwd[1024];

char fileMenuText[256];
char globalMenuText[256];
char functionMenuText[256];
char functionMenuTextShort[256];
char functionMenuTextLong[256];
char modifyMenuText[256];
char sortMenuText[256];

int viewMode = 0;

char sortmode[5] = "name";
char timestyle[9] = "locale";
int reverse = 0;
int human = 0;
int si = 0;
int ogavis = 3;
int ogapad = 1;
int showbackup = 1;
int colormode = 0;
int danger = 0;
int invalidstart = 0;
int filecolors = 0;
int markedinfo = 0;
int markedauto = 0;
int useEnvPager = 0;

int enterAsShow = 0;

int messageBreak = 0;

char *objectWild;

results *ob;

extern history *hs;
extern int topfileref;
extern int selected;
extern int totalfilecount;
extern char sortmode[5];
extern int showhidden;

extern char globalConfLocation[128];
extern char homeConfLocation[128];

extern char themeName[128];

struct sigaction sa;

extern int exitCode;

int setMarked(char* markedinput);
int checkStyle(char* styleinput);

void readConfig(const char * confFile)
{
  config_t cfg;
  config_setting_t *root, *setting, *group, *array; //probably don't need the array, but it may be used in the future.
  char markedParam[8];
  config_init(&cfg);
  if (config_read_file(&cfg, confFile)){
    // Deal with the globals first
    group = config_lookup(&cfg, "common");
    if (group){
      setting = config_setting_get_member(group, "theme");
      if (setting){
        if (!getenv("DFS_THEME_OVERRIDE")){
          strcpy(themeName, config_setting_get_string(setting));
          setenv("DFS_THEME", themeName, 1);
        }
      }
    }
    // Now for program specific
    group = config_lookup(&cfg, PROGRAM_NAME);
    if (group){
      // Check File Colour
      setting = config_setting_get_member(group, "color");
      if (setting){
        if (config_setting_get_int(setting)){
          filecolors = 1;
        }
      }
      // Check Marked
      setting = config_setting_get_member(group, "marked");
      if (setting){
        strcpy(markedParam, config_setting_get_string(setting));
        setMarked(markedParam);
      }
      // Check Sort
      setting = config_setting_get_member(group, "sortmode");
      if (setting){
        strcpy(sortmode, config_setting_get_string(setting));
      }
      // Check Reverse
      setting = config_setting_get_member(group, "reverse");
      if (setting){
        if (config_setting_get_int(setting)){
          reverse = 1;
        }
      }
      // Check Timestyle
      setting = config_setting_get_member(group, "timestyle");
      if (setting){
        strcpy(timestyle, config_setting_get_string(setting));
        if(!checkStyle(timestyle)){
          strcpy(timestyle, "locale");
        }
      }
      // Check Hidden
      setting = config_setting_get_member(group, "hidden");
      if (setting){
        if (config_setting_get_int(setting)){
          showhidden = 1;
        }
      }
      // Check Ignore Backups
      setting = config_setting_get_member(group, "ignore-backups");
      if (setting){
        if (config_setting_get_int(setting)){
          showbackup = 0;
        }
      }
      // Check No SF
      setting = config_setting_get_member(group, "no-sf");
      if (setting){
        if (config_setting_get_int(setting)){
          useEnvPager = 1;
        }
      }
      // Check No Danger
      setting = config_setting_get_member(group, "no-danger");
      if (setting){
        if (config_setting_get_int(setting)){
          danger = 0;
        }
      }
      // Check SI
      setting = config_setting_get_member(group, "si");
      if (setting){
        if (config_setting_get_int(setting)){
          si = 1;
        }
      }
      // Check Human Readable
      setting = config_setting_get_member(group, "human-readable");
      if (setting){
        if (config_setting_get_int(setting)){
          human = 1;
        }
      }
      // Check Enter As Show
      setting = config_setting_get_member(group, "show-on-enter");
      if (setting){
        if (config_setting_get_int(setting)){
          enterAsShow = 1;
        }
      }
    }
    // Check owner column
    group = config_lookup(&cfg, "show.owner");
    if (group){
      ogavis = 0;
      setting = config_setting_get_member(group, "owner");
      if (setting){
        if (config_setting_get_int(setting)){
          ogavis = ogavis + 1;
        }
      }
      setting = config_setting_get_member(group, "group");
      if (setting){
        if (config_setting_get_int(setting)){
          ogavis = ogavis + 2;
        }
      }
      setting = config_setting_get_member(group, "author");
      if (setting){
        if (config_setting_get_int(setting)){
          ogavis = ogavis + 4;
        }
      }
    }
  };
  config_destroy(&cfg);
}

int directory_view(char * currentpwd)
{
  objectWild = objectFromPath(currentpwd);
  if ( strchr(objectWild, MULTICHAR) || strchr(objectWild, ONECHAR)){
    strcpy(currentpwd, dirFromPath(currentpwd));
  } else {
    strcpy(objectWild, "");
  }

  topfileref = 0;
  clear();
  setColors(COMMAND_PAIR);

  // directory_top_menu();

  printMenu(0, 0, fileMenuText);

  set_history(currentpwd, "", "", 0, 0);
  free(ob);
  ob = get_dir(currentpwd);
  reorder_ob(ob, sortmode);
  display_dir(currentpwd, ob, topfileref, 0);

  // function_key_menu();

  printMenu(LINES-1, 0, functionMenuText);

  refresh();

  directory_view_menu_inputs();

  free(ob); //freeing memory
  return 0;
}

int global_menu()
{
  clear();

  printMenu(0, 0, globalMenuText);

  refresh();

  global_menu_inputs();

  return 0;
}

int setMarked(char* markedinput)
{
  int status = -1;
  if (!strcmp(markedinput, "always")){
    markedinfo = 1;
    markedauto = 0;
    status = 0;
  } else if (!strcmp(markedinput, "never")){
    markedinfo = 0;
    markedauto = 0;
    status = 0;
  } else if (!strcmp(markedinput, "auto")){
    markedinfo = 0;
    markedauto = 1;
    status = 0;
  }
  return(status);
}

int checkStyle(char* styleinput)
{
  if ( (strcmp(styleinput, "long-iso") && (strcmp(styleinput, "full-iso") && strcmp(styleinput, "iso") && strcmp(styleinput, "locale") ))){
      return 0;
    } else {
      return 1;
    }
}

int setColor(char* colorinput)
{
  if (!strcmp(colorinput, "always")){
    filecolors = 1;
    return 1;
    // } else if (!strcmp(colorinput, "auto")){
    //   filecolors = 0; // Need to make this autodetect
    //   return 1;
  } else if (!strcmp(colorinput, "never")){
    filecolors = 0;
    return 1;
  } else {
    return 0;
  }
}

void refreshScreen()
{
  endwin();
  clear();
  refresh();
  initscr();
  //mvprintw(0,0,"%d:%d", LINES, COLS);
  if (COLS < 89){
    strcpy(functionMenuText, functionMenuTextShort);
  } else {
    strcpy(functionMenuText, functionMenuTextLong);
  }
  switch(viewMode)
    {
    case 0:
      printMenu(0, 0, fileMenuText);
      printMenu(LINES-1, 0, functionMenuText);
      resizeDisplayDir(ob);
      break;
    case 1:
      printMenu(0, 0, globalMenuText);
      printMenu(LINES-1, 0, functionMenuText);
      resizeDisplayDir(ob);
      break;
    case 2:
      printMenu(0, 0, modifyMenuText);
      printMenu(LINES-1, 0, functionMenuText);
      resizeDisplayDir(ob);
      break;
    case 3:
      printMenu(0, 0, sortMenuText);
      printMenu(LINES-1, 0, functionMenuText);
      resizeDisplayDir(ob);
      break;
    case 4:
      printMenu(0, 0, globalMenuText);
      break;
    }
}

void sigwinchHandle(int sig){
  refreshScreen();
}


void printHelp(char* programName){
  printf (("Usage: %s [OPTION]... [FILE]...\n"), programName);
  fputs ((PROGRAM_DESC), stdout);
  fputs (("\n\
Sorts objects alphabetically if -St is not set.\n\
"), stdout);
  fputs (("\n\
Options shared with ls:\n\
  -a, --all                    do not ignore entries starting with .\n\
      --author                 prints the author of each file\n\
  -B, --ignore-backups         do not list implied entries ending with ~\n\
      --color[=WHEN]           colorize the output, see the color section below\n\
  -f                           do not sort, enables -aU\n\
  -g                           only show group\n\
  -G, --no-group               do not show group\n\
  -h, --human-readable         print sizes like 1K 234M 2G etc.\n\
      --si                     as above, but use powers of 1000 not 1024\n\
  -r, --reverse                reverse order while sorting\n\
  -S                           sort file by size, largest first\n\
      --time-style=TIME_STYLE  time/date format, see TIME_STYLE section below\n\
  -t                           sort by modification time, newest first\n\
  -U                           do not sort; lists objects in directory order\n\
      --help                   displays help message, then exits\n\
      --version                displays version, then exits\n"), stdout);
  fputs (("\n\
The TIME_STYLE argument can be: full-iso; long-iso; iso; locale.\n"), stdout);
  fputs (("\n\
Using color to highlight file attributes is disabled by default and with\n\
--color=never. With --color or --color=always this function is enabled.\n"), stdout);
  fputs (("\n\
Options specific to show:\n\
      --theme=[THEME]          color themes, see the THEME section below for\n\
                               valid themes\n\
      --no-danger              turns off danger colors when running with\n\
                               elevated privileges\n\
      --marked=[MARKED]        shows information about marked objects. See\n\
                               MARKED section below for valid options\n\
      --no-sf                  does not display files in sf\n\
      --show-on-enter          repurposes the Enter key to launch the show\n\
                               command\n"), stdout);
  fputs (("\n\
The THEME argument can be:\n"), stdout);
  listThemes();
  fputs (("\n\
Exit status:\n\
 0  if OK,\n\
 1  if minor problems (e.g., cannot access subdirectory),\n\
 2  if major problems (e.g., cannot access command-line arguement).\n"), stdout);
  printf ("\nPlease report any bugs to: <%s>\n", PACKAGE_BUGREPORT);
}

int main(int argc, char *argv[])
{
  uid_t uid=getuid(), euid=geteuid();
  int c;

  // Set Config locations
  setConfLocations();

  // Check if we're root to display danger
  if (uid == 0 || euid == 0){
    danger = 1;
  }

  // Read the config

  readConfig(globalConfLocation);
  readConfig(homeConfLocation);

  // Check for theme env variable
  if ( getenv("DFS_THEME")) {
    strcpy(themeName, getenv("DFS_THEME"));
  }

  // Getting arguments
  while (1)
    {
      static struct option long_options[] =
        {
         {"all",            no_argument,       0, 'a'},
         {"almost-all",     no_argument,       0, 'A'},
         {"author",         no_argument,       0, GETOPT_AUTHOR_CHAR},
         {"ignore-backups", no_argument,       0, 'B'},
         {"human-readable", no_argument,       0, 'h'},
         {"no-group",       no_argument,       0, 'G'},
         {"reverse",        no_argument,       0, 'r'},
         {"time-style",     required_argument, 0, GETOPT_TIMESTYLE_CHAR},
         {"si",             no_argument,       0, GETOPT_SI_CHAR},
         {"help",           no_argument,       0, GETOPT_HELP_CHAR},
         {"version",        no_argument,       0, GETOPT_VERSION_CHAR},
         {"no-danger",      no_argument,       0, GETOPT_NODANGER_CHAR},
         {"color",          optional_argument, 0, GETOPT_COLOR_CHAR},
         {"theme",          optional_argument, 0, GETOPT_THEME_CHAR},
         {"marked",         optional_argument, 0, GETOPT_MARKED_CHAR},
         {"no-sf",          no_argument,       0, GETOPT_ENVPAGER_CHAR},
         {"show-on-enter",  no_argument,       0, GETOPT_SHOWONENTER_CHAR},
         {0, 0, 0, 0}
        };
      int option_index = 0;

      c = getopt_long(argc, argv, "aABfgGhrStU", long_options, &option_index);

      if ( c == -1 ){
        break;
      }

    switch(c){
    case 'A':
      // Dropthourgh
    case 'a':
      showhidden = 1;
      break;
    case GETOPT_AUTHOR_CHAR:
      ogavis = ogavis + 4;
      break;
    case 'B':
      showbackup = 0;
      break;
    case GETOPT_COLOR_CHAR:
      if (optarg){
        if (!setColor(optarg)){
          printf("%s: invalid argument '%s' for 'color'\n", argv[0], optarg);
          fputs (("\
Valid arguments are:\n\
  - always\n\
  - never\n"), stdout);
          printf("Try '%s --help' for more information.\n", argv[0]);
          exit(2);
        }
      } else {
        filecolors = 1;
        }
      break;
    case GETOPT_THEME_CHAR:
      if (optarg){
        if (strcmp(optarg, "\0")){
          strcpy(themeName, optarg);
          setenv("DFS_THEME_OVERRIDE", "TRUE", 1);
        }
      } else {
        printf("%s: The following themes are available:\n", argv[0]);
        listThemes();
        exit(2);
      }
      break;
    case 'f':
      strcpy(sortmode, "none"); // This can be set to anything non valid
      showhidden = 1;
      break;
    case 'S':
      strcpy(sortmode, "size");
      break;
    case GETOPT_TIMESTYLE_CHAR:
      strcpy(timestyle, optarg);
      if (!checkStyle(timestyle)){
        printf("%s: invalid argument '%s' for 'time style'\n", argv[0], timestyle);
        fputs (("\
Valid arguments are:\n\
  - full-iso\n\
  - long-iso\n\
  - iso\n\
  - locale\n"), stdout);
        printf("Try '%s --help' for more information.\n", argv[0]);
        exit(2);
      }
      break;
    case 't':
      strcpy(sortmode, "date");
      break;
    case 'g':
      ogavis--;
      break;
    case 'G':
      ogavis = ogavis - 2;
      break;
    case 'h':
      human = 1;
      break;
    case GETOPT_SI_CHAR:
      human = 1;
      si = 1;
      break;
    case 'r':
      reverse = 1;
      break;
    case 'U':
      strcpy(sortmode, "none"); // Again, invalid
      break;
    case GETOPT_NODANGER_CHAR:
      danger = 0;
      break;
    case GETOPT_MARKED_CHAR:
      if (optarg){
        if ( setMarked(optarg) == -1 ){
          printf("%s: invalid argument '%s' for 'marked'\n", argv[0], optarg);
          fputs (("\
Valid arguments are:\n\
  - always\n\
  - never\n\
  - auto\n"), stdout);
          printf("Try '%s --help' for more information.\n", argv[0]);
          exit(2);
        }
      }
      break;
    case GETOPT_HELP_CHAR:
      printHelp(argv[0]);
      exit(0);
      break;
    case GETOPT_VERSION_CHAR:
      printVersion(PROGRAM_NAME);
      exit(0);
      break;
    case GETOPT_ENVPAGER_CHAR:
      useEnvPager = 1;
      break;
    case GETOPT_SHOWONENTER_CHAR:
      enterAsShow = 1;
      break;
    default:
      // abort();
      exit(2);
    }
  }


  // If all author is also requested, the padding needs adjusting.
  if ( ogavis == 7 ){
    ogapad = 2;
  }


  // Writing Menus
  strcpy(fileMenuText, "!Copy, !Delete, !Edit, !Hidden, !Modify, !Quit, !Rename, !Show, h!Unt, e!Xec");
  strcpy(globalMenuText, "c!Olors, !Run command, !Edit file, !Help, !Make dir, !Quit, !Show dir");
  strcpy(functionMenuTextShort, "<F1>-Down <F2>-Up <F3>-Top <F4>-Bottom <F5>-Refresh <F6>-Mark/Unmark <F7>-All <F8>-None <F9>-Sort");
  strcpy(functionMenuTextLong, "<F1>-Down <F2>-Up <F3>-Top <F4>-Bottom <F5>-Refresh <F6>-Mark/Unmark <F7>-All <F8>-None <F9>-Sort <F10>-Block");
  strcpy(modifyMenuText, "Modify: !Owner/Group, !Permissions");
  strcpy(sortMenuText, "Sort list by - !Date & time, !Name, !Size");

  set_escdelay(10);
  //ESCDELAY = 10;

  setlocale(LC_ALL, "");

  initscr();

  // Decide which function menu needs initially printing
  if (COLS < 89){
    strcpy(functionMenuText, functionMenuTextShort);
  } else {
    strcpy(functionMenuText, functionMenuTextLong);
  }

  memset(&sa, 0, sizeof(struct sigaction));
  sa.sa_handler = sigwinchHandle;
  sigaction(SIGWINCH, &sa, NULL);

  start_color();
  cbreak(); //Added for new method
  setDefaultTheme();
  loadAppTheme(themeName);
  bkgd(COLOR_PAIR(DISPLAY_PAIR));
  cbreak();
  // nodelay(stdscr, TRUE);
  noecho();
  curs_set(FALSE); // Hide Curser (Will want to bring it back later)
  keypad(stdscr, TRUE);



  // Remaining arguments passed as working directory
  if (optind < argc){
    if (!check_first_char(argv[optind], "/")){
      // If the path given doesn't start with a / then assume we're dealing with a relative path.
      getcwd(currentpwd, sizeof(currentpwd));
      sprintf(currentpwd, "%s/%s", currentpwd, argv[optind]);
    } else {
      strcpy(currentpwd, argv[optind]);
    }
    chdir(currentpwd);
  } else {
    getcwd(currentpwd, sizeof(currentpwd));
  }

  if (!check_dir(currentpwd)){
    //strcpy(currentpwd, "/"); // If dir doesn't exist, default to root
    invalidstart = 1;
    exitCode = 1;
    global_menu();
  }
  testSlash:
  if (check_last_char(currentpwd, "/") && strcmp(currentpwd, "/")){
    currentpwd[strlen(currentpwd) - 1] = '\0';
    goto testSlash;
  }
  directory_view(currentpwd);
  return exitCode;
}
